Implement authorization layer using pundit

ankitkataria opened the pull request:
Fixes #150

Fixes #240 

#### Describe the changes you have made in this pr -

Add pundit
Working on adding policies for the following controllers -
- [X] assignment
- [X] group_members
- [X] group
- [X] projects
- [X] simulator
- ~logix~ (not required)
- ~star~ (not required)
- ~collaboration~ (not required)

ankitkataria commented:
@tachyons Please review! :gift: 
I've added the required policies. Currently, I've committed the changes relating to a single policy as one commit so that it is easier to single out an error in auth policies to a single commit, if it ever occurs later. Let me know if they are required to be squashed!
Thanks. :smile:

satu0king commented:
Looks good! After making the suggested changes by @tachyons I will merge this PR. This is a lot of changes and deals with security/logic of the code and therefore has to be thoroughly tested, so I request the mentors to not merge this PR immediately.

tachyons commented:
That is unlikely, rails admin should work with pundit as well. Can you please confirm ?

ankitkataria commented:
Okay, I'll look it up in depth once.
Also starting to work on adding tests.

On Wed 6 Mar, 2019, 4:48 PM Aboobacker MK, <notifications@github.com> wrote:

> That is unlikely, rails admin should work with pundit as well. Can you
> please confirm ?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/CircuitVerse/CircuitVerse/pull/184#issuecomment-470070257>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AVAEHgdYUbzqYKqLkxWdEunM9Hb9_if1ks5vT6QIgaJpZM4bbV6T>
> .
>

ankitkataria commented:
@tachyons Removed cancan. To configure rails admin I used [this](https://github.com/CircuitVerse/CircuitVerse/pull/184/files#diff-789ebe8a87cb636c99034bb438302d7eR18) hack. Saved me from adding a dependency that was necessary for pundit to be used with rails_admin. Also added tests for all policies! 
@satu0king Fixed that error in rake task in this pr. Sorry about that, won't happen again :sweat:

satu0king commented:
> @tachyons Removed cancan. To configure rails admin I used [this](https://github.com/CircuitVerse/CircuitVerse/pull/184/files#diff-789ebe8a87cb636c99034bb438302d7eR18) hack. Saved me from adding a dependency that was necessary for pundit to be used with rails_admin. Also added tests for all policies!
> @satu0king Fixed that error in rake task in this pr. Sorry about that, won't happen again 😓

No need to apologize @ankitkataria !

ankitkataria commented:
@tachyons should we remove the last commit or keep it for backward compatibility?

tachyons commented:
No need to remove it

ankitkataria commented:
@tachyons Has this been tested and reviewed?

tachyons commented:
https://github.com/CircuitVerse/CircuitVerse/pull/184#issuecomment-469978422 Waiting for @satu0king

ankitkataria commented:
@satu0king Please review!

danllin commented:
#150

satu0king commented:
Hey! Pundit looks good, I was looking for something like this during the initial stages of development. 

The implementation looks good. I ran into the following error when I started testing it. I made a project with public access, and in incognito (signed out user) I tried accessing the page. 

![image](https://user-images.githubusercontent.com/2092958/54416803-6f1cd280-4726-11e9-8614-0666d22dab1e.png)



# Please ensure the following scinarios are handled. 
1. User logged in authorized
1. user logged in unauthorized -> Show unauthorized
1. user not logged in -> Request login 

**Admin should be able to access any project/assignment/group irrespective of the settings**

## Projects Policy 
* Public/Limited Access -> Anyone can see, only authors and collaborators can edit
* Private -> Only authors and collaborators can edit/see
* Only author can add members as collaboration. 

## Group - Assignment 
* Only mentor and members of the group can see any group or assignment 
* Only mentor can edit the assignment, only mentor can see all the assignment projects
* Only admin can add members to the group

I might have missed some corner cases, so please take care.

tachyons commented:
When did we decided to use double quotes over single quotes :thinking:

ankitkataria commented on lib/tasks/spec.rake:
```diff
@@ -1,5 +1,7 @@
-  require 'rspec/core/rake_task'
-  RSpec::Core::RakeTask.new(:all)
-rescue LoadError
+  begin
```
Was asked to keep this change for backward compatibility

tachyons replied:
You can disable a particular cop for a line using ` # rubocop-disable <Cop name> `

ankitkataria commented:
@tachyons For string interpolations(like `puts "Hey #{name}"`), we required double quotes. I felt using double quotes everywhere made the standards uniform. Do we want to remove this?

tachyons commented:
It is actually a controversial topic in ruby community and both side have their own reasoning. Both are fine as long as we are consistant and stick to either of these

ankitkataria commented:
@tachyons Thanks for the comments! I've fixed them. The trick with rubocop was really neat! :smile:

ankitkataria commented:
@satu0king Thanks for the elaborate review! :smile: 
I've fixed the error for public projects. Also fixed Issue #240 in this one. 
There were a few admin accesses missing at some places. I've added those. 
Also, I've manually tested (phew..) everything for the cases that you've mentioned. As for now admin can't access the groups page of individual users from `users/logix::groups` because of #256. I can fix it once this is merged.

satu0king commented on app/controllers/assignments_controller.rb:
```diff
@@ -37,20 +34,18 @@ def start
+    @assignment.group_id = params[:group_id]
     @assignment.deadline = Time.now+1.week
+    authorize @assignment, :admin_access?
   end
```
I might have got this wrong, why not have `authorize @assignment, :admin_access?` as a before action hook? just like edit, update, destroy and reopen.

ankitkataria replied:
Here I need an `@assignment` object before doing an authorize, in the new method `@assignment` is not already set like in edit, update, destroy and reopen because it doesn't exist in the database yet. I have to still create it. So I have to create a dummy object like so `@assignment = Assignment.new` and assign the group_id to do the respective authorize.

satu0king replied:
Okay

satu0king commented on app/controllers/assignments_controller.rb:
```diff
@@ -37,20 +34,18 @@ def start
-    if @assignment.status == 'open'
-      render plain: "already open" and return;
-    end
+    authorize @assignment
```
Again, I might be wrong here, we are already checking for admin access so why do need to authorize again?

ankitkataria replied:
See `assignment_policy` file, `reopen?` has another check to see if the assignment isn't already open and it raises its own custom error [here](https://github.com/CircuitVerse/CircuitVerse/pull/184/files#diff-c466b4128cdaa10a4f59eb1cb6be506cR30) That is why it is needed here again.

satu0king replied:
Yes, I realized you have moved some of the non-authorization logic to the policies

satu0king commented on app/controllers/collaborations_controller.rb:
```diff
@@ -30,10 +34,7 @@ def create
-    if(@project.author_id!=current_user.id)
-      render plain: "Access Restricted " and return
-    end
+    authorize @project, :author_access?
```
I might have got this wrong, why not have authorize @assignment, :admin_access? as a before action hook? just like edit, update, destroy and reopen.

satu0king commented:
@ankitkataria, I tested everything, everything looks good. I am curious as to how the user is initialized.

satu0king commented:
Amazing PR!

ankitkataria commented:
@satu0king The user is automatically initialized by `Pundit` internally from `current_user`. If the user is logged in then it returns a user object else returns nil.
